File: airesources/Java/hlt/GameMap.java
package hlt;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;
import java.util.Collections;
import java.util.Collection;

public class GameMap {
    private final int width, height;
    private final int playerId;
    private final List<Player> players;
    private final List<Player> playersUnmodifiable;
    private final Map<Integer, Planet> planets;
    private final List<Ship> allShips;
    private final List<Ship> allShipsUnmodifiable;

    // used only during parsing to reduce memory allocations
    private final List<Ship> currentShips = new ArrayList<>();

    public GameMap(final int width, final int height, final int playerId) {
        this.width = width;
        this.height = height;
        this.playerId = playerId;
        players = new ArrayList<>(Constants.MAX_PLAYERS);
        playersUnmodifiable = Collections.unmodifiableList(players);
        planets = new TreeMap<>();
        allShips = new ArrayList<>();
        allShipsUnmodifiable = Collections.unmodifiableList(allShips);
    }

    public int getHeight() {
        return height;
    }

    public int getWidth() {
        return width;
    }

    public int getMyPlayerId() {
        return playerId;
    }

    public List<Player> getAllPlayers() {
        return playersUnmodifiable;
    }

    public Player getMyPlayer() {
        return getAllPlayers().get(getMyPlayerId());
    }

    public Ship getShip(final int playerId, final int entityId) throws IndexOutOfBoundsException {
        return players.get(playerId).getShip(entityId);
    }

    public Planet getPlanet(final int entityId) {
        return planets.get(entityId);
    }

    public Map<Integer, Planet> getAllPlanets() {
        return planets;
    }

    public List<Ship> getAllShips() {
        return allShipsUnmodifiable;
    }

    public ArrayList<Entity> objectsBetween(Position start, Position target) {
        final ArrayList<Entity> entitiesFound = new ArrayList<>();

        addEntitiesBetween(entitiesFound, start, target, planets.values());
        addEntitiesBetween(entitiesFound, start, target, allShips);

        return entitiesFound;
    }

    private static void addEntitiesBetween(final List<Entity> entitiesFound,
                                           final Position start, final Position target,
                                           final Collection<? extends Entity> entitiesToCheck) {

        for (final Entity entity : entitiesToCheck) {
            if (entity.equals(start) || entity.equals(target)) {
                continue;
            }
            if (Collision.segmentCircleIntersect(start, target, entity, Constants.FORECAST_FUDGE_FACTOR)) {
                entitiesFound.add(entity);
            }
        }
    }

    public Map<Double, Entity> nearbyEntitiesByDistance(final Entity entity) {
        final Map<Double, Entity> entityByDistance = new TreeMap<>();

        for (final Planet planet : planets.values()) {
            if (planet.equals(entity)) {
                continue;
            }
            entityByDistance.put(entity.getDistanceTo(planet), planet);
        }

        for (final Ship ship : allShips) {
            if (ship.equals(entity)) {
                continue;
            }
            entityByDistance.put(entity.getDistanceTo(ship), ship);
        }

        return entityByDistance;
    }

    public GameMap updateMap(final Metadata mapMetadata) {
        final int numberOfPlayers = MetadataParser.parsePlayerNum(mapMetadata);

        players.clear();
        planets.clear();
        allShips.clear();

        // update players info
        for (int i = 0; i < numberOfPlayers; ++i) {
            currentShips.clear();
            final Map<Integer, Ship> currentPlayerShips = new TreeMap<>();
            final int playerId = MetadataParser.parsePlayerId(mapMetadata);

            final Player currentPlayer = new Player(playerId, currentPlayerShips);
            MetadataParser.populateShipList(currentShips, playerId, mapMetadata);
            allShips.addAll(currentShips);

            for (final Ship ship : currentShips) {
                currentPlayerShips.put(ship.getId(), ship);
            }
            players.add(currentPlayer);
        }

        final int numberOfPlanets = Integer.parseInt(mapMetadata.pop());

        for (int i = 0; i < numberOfPlanets; ++i) {
            final List<Integer> dockedShips = new ArrayList<>();
            final Planet planet = MetadataParser.newPlanetFromMetadata(dockedShips, mapMetadata);
            planets.put(planet.getId(), planet);
        }

        if (!mapMetadata.isEmpty()) {
            throw new IllegalStateException("Failed to parse data from Halite game engine. Please contact maintainers.");
        }

        return this;
    }
}

File: airesources/Scala/hlt/GameMap.scala
package hlt

object GameMap {
  private def addEntitiesBetween(start: Position,
                                 target: Position,
                                 entitiesToCheck: Iterable[Entity]): Iterable[Entity] = {
    entitiesToCheck
      .filter(_ != start)
      .filter(_ != target)
      .filter(entity =>
        Collision.segmentCircleIntersect(start, target, entity, Constants.FORECAST_FUDGE_FACTOR))
  }
}

class GameMap(val width: Short,
              val height: Short,
              val playerId: Short,
              val metadataIterator: Iterator[String]) {
  DebugLog.addLog("--- NEW TURN ---")
  var players: Seq[Player] = {
    val numberOfPlayers = MetadataParser.parsePlayerNum(metadataIterator)

    // update players info
    def getPlayerFromMetadata: Player = {
      val playerId = MetadataParser.parsePlayerId(metadataIterator)
      val ships = MetadataParser.getShipList(Some(playerId), metadataIterator)
      new Player(playerId, ships.map(ships => ships.id -> ships).toMap)
    }

    for (counter <- 1 to numberOfPlayers) yield getPlayerFromMetadata
  }
  DebugLog.addLog("Players: " + players)

  var planets: Map[Long, Planet] = {
    val numberOfPlanets = metadataIterator.next.toLong
    (for (counter <- 1 to numberOfPlanets.toInt)
      yield MetadataParser.newPlanetFromMetadata(metadataIterator))
      .map(planet => planet.id -> planet)
      .toMap
  }
  DebugLog.addLog("Planets: " + planets)

  if (!metadataIterator.isEmpty) {
    throw new IllegalStateException(
      "Failed to parse data from Halite game engine. Please contact maintainers.")
  }

  var allShips: Seq[Ship] = {
    def getShipsFromPlayers(players: Seq[Player], shipAccum: Seq[Ship]): Seq[Ship] = {
      if (players.isEmpty) {
        shipAccum
      } else {
        getShipsFromPlayers(players.tail, players.head.getShips.values.toSeq ++ shipAccum)
      }
    }

    getShipsFromPlayers(players, Seq.empty)
  }
  DebugLog.addLog("All ships: " + allShips)

  def getAllPlayers: Seq[Player] = players

  def getMyPlayer: Player = players(getMyPlayerId)

  def getMyPlayerId: Short = playerId

  @throws[IndexOutOfBoundsException]
  def getShip(playerId: Short, entityId: Long): Option[Ship] =
    players(playerId).getShip(entityId)

  def getPlanet(entityId: Long): Planet = planets(entityId)

  def getAllPlanets: Map[Long, Planet] = planets

  def getAllShips: Seq[Ship] = allShips

  def objectsBetween(start: Position, target: Position): Iterable[Entity] = {
    GameMap.addEntitiesBetween(start, target, planets.values) ++ GameMap
      .addEntitiesBetween(start, target, allShips)
  }

  def nearbyEntitiesByDistance(entity: Entity): Map[Double, Entity] = {
    (planets.values
      .filter(_ != entity)
      .map(planet => entity.getDistanceTo(planet) -> planet)
      .toMap[Double, Entity]
      ++
        allShips
          .filter(_ != entity)
          .map(planet => entity.getDistanceTo(planet) -> planet)
          .toMap[Double, Entity])
  }
}
